README.md
# SpectrWm-Binary-Build-Github-Readme.MD
      ___ ___  ___ ____/ /______    ____ _  
     (_-</ _ \/ -_) __/ __/ __/ |/|/ /  ' \ 
    /___/ .__/\__/\__/\__/_/  |__,__/_/_/_/ 
        /_/                                 

     _   _       _        _             
    |_) |_  /\  | \ |\/| |_    _| ._ _  
    | \ |_ /--\ |_/ |  | |_ o (_| | | | 

## Overview:
This SpectrWm Manual Binary Build & Manual Install was created so Future Updates of the binary can be easily implement and with the use of symlinks,
to my ~/bin where I keep all my scripts files. I want to upgrade spectrwm without the use of the Disro's Repo's. and ease of upgrading my "your" spectrwm install.
All you have to do is build the new binary file and place it in ~/bin (In My Build)...

#### This project is for & tested on Linux Mint 20. While it may work on other Distros this has NOT BEEN TESTED with Other Distros...

## Known Issues:
#### I have had the "Make" Fail 2 times. When or IF this happens, Spectrwm will not load. To check this, go to your ~/bin folder in a or open a terminal in that folder and type specrtwm on the command line (CLI) you should get a output that states you have spectrwm 3.4.1 and another WM is running. If you do not get this output then the build had Failed... Just "Make it again, copy it to your ~/bin folder and chmod +x the file and it should now load. No guarantee this will work...

## Inspiration:
This project was the result of a video that The Linux Dabbler produced on 6/29/2020, the video.(https://www.youtube.com/watch?v=tYC7qxnjka4}
explained how to build the binaries and what dependencies might be needed of the build of SpectrWm.
The whole point of the video was to show how to install SpectWm without using the outdated Repo versions and use the binaries from SpectrWm's GitHub page
(https://github.com/conformal/spectrwm) and how to install it on a headless OS install where the systems boots straight in to SpectrWm.
This was a great video and got me to think, can it be done on Linux Mint 20 in a VM using LightDm login manager...

## Goals:
My goal of this project are to see if SpectrWm can be build, installed, and make it so it can be loaded using LightDm Login Manager and make the build without
using "make install" to get everything working...
Now after I spent 18 ours exploring this project, installing and testing all the dependencies needed to make this happen. I seen in the comments of
said video from Damiel Olsson that suggested another way to do the dependencies easier... What luck... :-) Just my Luck... Lol

........
Daniel Olsson said;
Nice video, sometimes I use sudo apt build-dep ( package name) in this case spectrwm, then it will automatically pull in everything you need to recompile
SpectrWm. :-)
........

While I have not tested this to see, I like the idea...

## Looked at the above suggestion...
I did looked into this on Linux Mint 20 and it would require making changes to the source.list (which is blank) and add a bunch of stuff add to the file and formatted
a curtain way and so on, so I elected to use my own Dependencies script instead of this method...

I have added my bash script I made and all the files and dependencies I used to make this project work for ME and how I got Linux Mint 20 to show
SpectrWm in the LightDm Login Menu and load SpectrWm window manager...

## Files of the project:
 Install-guide.md
 The install instruction

 build-depend-mint20.sh
 The dependencies Bash install script

 After-Build.sh
 The Bash script to finish the install

## Special Files:
I have created "build-depend-mint20.sh" this will install the dependencies!
I have made a Bash file called "After-Build.sh" to finish the install process. The "After-Build.sh" will copy all the files to their proper places and set the file
permissions, symlinks and make the directories that are needed and put the spectrwn.config in the proper place in the USER .config folder..

------------------------------------------------------

Here are some of the NOTES from the Bash Script...

Script Date: 06/29/2020, Script Creator: Bruce E. Scott, Github: https://github.com/bescott9944, Forums: https://www.ezeelinux.com/talk/bescott9944,
https://forum.endeavouros.com/bescott9944, licensed under the GNU and MIT respectively

PURPOSE Of This Script:

This is install guide and process I followed to build the binary files for installing/building the new Spectrwm build and the dependencies that are needed to build the newer
binary version of SpectrWm... I am doing this build to make a manual install... 06/29/2020 -Bruce

This script and project has been inspired by a video from The Linux Dabbler YouTube Channel. His Channel: https://www.youtube.com/channel/UC3pasXoRUTKPpQaKtlJz62g,
The Video: https://www.youtube.com/watch?v=tYC7qxnjka4

NOTE: This Bash file has been limited to 80 characters. You may need edit the line and change the "next line break ( \ )" of the bash file to suite your needs!

This Install dependencies list has been tested on Linux Mint 20, several times. 06/29/2020

Note: Some dependencies require the " -dev " versions of the dependencies and other do not. When dependencies are installed they will intern install their own dependencies too..
I don't know if this is do to Mint/Ubuntu Repos or not.

The do the build I needed all these dependencies before the "make" part of the build would run with out any error in the output of the "make" and before the binary file would
compile with out any errors.

There was a lot of trial and error and a few hours of testing to produce this list. You mileage may vary as dependencies and files change over time. with that said this script is
a good starting place. I have not tested this script on anything other than Linux Mint 20, in a VM for testing purpose...

#### WARNING: USE AT YOUR OWN RISK!

With that said, Myself or any of my affiliates will not be held responsible for any loss of data, software, operating systems, damage to, wiping of, altering of, and any other thing that might happen using this script!That also includes the urges to drink lots of coffee, Rum, eating of junk food, and eating lots of Hot Pockets, and not limited to the building of bonfires, dancing around said bonfires and howling at the MOON!
#### YOU HAVE BEEN WARNED! -BES




Lead Developer: Bruce E. Scott, Script Creator: Bruce E. Scott.

Github: https://github.com/bescott9944.

Forums: https://www.ezeelinux.com/talk/bescott9944, https://forum.endeavouros.com/bescott9944.

Licensed under the GNU and MIT respectively



ChangeLog.MD
July 1st 2020
Created the Github page and setup all the files.
Added the Bash files, README.MD, Install.MD and so on.

July 2nd 2020
Added some more and updated the instructions;

July 4th 2020
I did another VM install and made some changes to the dependencies
and made a few additions to the instructions and the build process's.
I also did another install with the New Linux Mint 20 (Not The Beta)
in a VM.
Updated all the files a bit...


after-build.sh
#!/bin/bash
#
#      ___ ___  ___ ____/ /______    ____ _  
#     (_-</ _ \/ -_) __/ __/ __/ |/|/ /  ' \ 
#    /___/ .__/\__/\__/\__/_/  |__,__/_/_/_/ 
#        /_/                                 
#         _                _                       
#    /\ _|_ _|_  _  ._ __ |_)     o |  _|    _ |_  
#   /--\ |   |_ (/_ |     |_) |_| | | (_| o _> | | 
#
#-------------------------------------------------------
# Script build Date... 7/07/2020 By Bruce E. Scott
#
# This Script is to be run after the building of the binary of SpectrWm
# to transfer the "Make" files to the Directories to where they belong. Then it creat's
# the symlinks, directories, and set the permissions needed for everthing to work.

#
# Setting ERROR Function
 set -e

# Bin Function
# Making the USER bin like this ~/bin or /home/$USER/bin
if [[ -d /home/$USER/bin ]]; then

    cp /home/$USER/work/spectrwm-3.4.1/linux/spectrwm /home/$USER/bin && cp /home/$USER/work/spectrwm-3.4.1/linux/baraction.sh /home/$USER/bin

echo -e '\tUser ~/bin Directory Excises!\n \tCopying Spectrwm and baraction.sh Files'

# Making Spectrwm and baraction.sh extricable in ~/bin folder.
   chmod +rwx /home/$USER/bin/spectrwm

   chmod +rwx /home/$USER/bin/baraction.sh

echo -e  '\tAll Files Permissions are Set' # <--- Fun Message

 else
    mkdir -p /home/$USER/bin && cp /home/$USER/work/spectrwm-3.4.1/linux/spectrwm /home/$USER/bin && cp /home/$USER/work/spectrwm-3.4.1/linux/baraction.sh /home/$USER/bin
##    mkdir -p /home/$USER/bin && cp /home/$USER/test* /home/$USER/bin/ # <---- This paths will need to be changed
echo -e '\tMade the ~/Bin Directory\n \thave Copied Spectrwm and Baraction.sh Files'

# Making Spectrwm nd baraction.sh extricable in ~/bin folder.
   chmod +rwx /home/$USER/bin/spectrwm

   chmod +rwx /home/$USER/bin/baraction.sh

echo -e '\tAll Files Permissions Set' # <--- Fun Message

fi

# Testing for xsessions dir and copying spectrWm.desktop to /usr/share/xsessions
 if [[ -d /usr/share/xsessions ]]  && [[ -f /home/$USER/work/spectrwm-3.4.1/linux/spectrwm.desktop ]]; then

    echo -e 'ENTERING Sudo Mode!\n \tYou Need to Enter Your Password to Proceed coping the Files!'
 sudo  cp /home/$USER/work/spectrwm-3.4.1/linux/spectrwm.desktop /usr/share/xsessions

 echo -e '\tCopied spectrWm.desktop! To /usr/share/xsessions Done!' # <--- Fun Message

 fi

# Checking for libswmhack.so.0.0 and copying it to /usr/local/lib
 if [[ -d /usr/local/lib ]] && [[ /home/$USER/work/spectrwm-3.4.1/linux/libswmhack.so.0.0 ]]; then

 sudo  cp -r /home/$USER/work/spectrwm-3.4.1/linux/libswmhack.so.0.0 /usr/local/lib

 echo -e '\tCopied libswmhack.so.0.0 to /usr/local/lib Done!' # <--- Fun Message

 fi

 if [[ -d /bin ]]
  then
 sudo ln -sf /home/$USER/bin/spectrwm /bin

 echo -e '\tCreated the Symlink in /bin Done!' # <--- Fun Message

 fi

 if [[ -d /home/$USER/work/spectrwm-3.4.1 ]] && [[ -f /home/$USER/work/spectrwm-3.4.1/spectrwm.conf ]]; then

    mkdir -p /home/$USER/.config/spectrwm && cp /home/$USER/work/spectrwm-3.4.1/spectrwm.conf /home/$USER/.config/spectrwm

 echo -e '\tCopied spectrwm binary to /home/USER/.config/spectrwm Done!' # <--- Fun Message

    else
        echo 'spectrwm.conf Not Found'
 fi

# END




spectrwn-build-depend-mint20.sh
# !/BIN/BASH
#
#
#      ___ ___  ___ ____/ /______    ____ _  
#     (_-</ _ \/ -_) __/ __/ __/ |/|/ /  ' \ 
#    /___/ .__/\__/\__/\__/_/  |__,__/_/_/_/ 
#        /_/                                 
#   _ ._   _   _ _|_ ._     ._ __ |_      o |  _| __ _|  _  ._   _  ._   _| __ ._ _  o ._ _|_  ) / \    _ |_  
#  _> |_) (/_ (_  |_ | \/\/ | |   |_) |_| | | (_|   (_| (/_ |_) (/_ | | (_|    | | | | | | |_ /_ \_/ o _> | | 
#     |                                                     |                                                 
#
# Script Date: 06/29/2020
# Script Creator: Bruce E. Scott
# Github: https://github.com/bescott9944
# Forums: https://www.ezeelinux.com/talk/bescott9944
#         https://forum.endeavouros.com/bescott9944
# Licensed under the GNU and MIT respectively
#
# PURPOSE Of This Script:
# This Script is for installing/building the new Spectrwm build and the
# dependencies that are needed to build the newer binary version
# of SpectrWm... Doing this build to make a manual install... 06/29/2020 -Bruce
# This script was inspired by a video from The Linux Dabbler YouTube Channel.
# His Channel: https://www.youtube.com/channel/UC3pasXoRUTKPpQaKtlJz62g
# The Video: https://www.youtube.com/watch?v=tYC7qxnjka4
#
# Change Log.
# 07/04/2020
# Made some changes to the dependencies list to work better. -Bruce
#
# NOTE: This Bash file has been limited to 80 characters. You may need edit
# the line and change the "next line break ( \ )" of the bash file to suite
# your needs!
#
# This Install dependencies list has been tested on Mint 20, 06/29/2020
#
# The otherday in a comment in said video, Daniel Olsson said:
# Nice video, sometimes i use sudo apt build-dep "packagename" in this case
# spectrwm, then it will automatically pull in everything you need to
# recompile spectrvm :)
# Wish I known this a few days ago... Lol
#
# Note: Some dependencies require the " -dev " versions of the dependencies and
# other do not. When dependencies are installed they will intern install their
# own dependencies too.. I don't know if this is do to Mint/Ubuntu Repos or not.
# The do the build I needed all these dependencies before the "make" part of the build
# would run with out any error in the output of the "make" and before the binary file
# would compile with out any errors In Linux Mint 20.
#
# The build that I did needed all these dependencies before the "make" part of
# the build would run with out any error in the output and before the binary
# file would compile.
# There was a lot of trial and error an a few hours of testing to produce this
# list. You mileage may vary as dependencies and files change over time.
# with that said this script is a good starting place.
# I have not tested this script on anything other than Linux Mint 20, in a VM
# for testing purpose...
#
# WARNING: USE AT YOUR OWN RISK!
# I or any of my affiliates will or will not be held responsible for any loss
# of data, software, operating systems, damage to, wiping of, altering of, and
# any other thing else that might happen using this script!
# That also includes the urges to drink lots of coffee, Rum, eating of
# junk food, and eating lots of Hot Pockets, and not to limited to the
# building of bonfires, dancing around said bonfires and howling at the MOON!
# YOU HAVE BEEN WARNED! -BES
# sudo apt install g++-multilib g++-9-multilib stterm xfonts-terminus-oblique
# xfonts-cyrillic


sudo apt install suckless-tools xterm libxt6 libc6 libx11-6 libx11-xcb1 \
libx11-xcb-dev libxcb-icccm4 libxcb-icccm4-dev libxcb-keysyms1 \
libxcb-keysyms1-dev libxcb-randr0 \
libxcb-randr0-dev libxcb-util1 libxcb-xinput0 libxcb-xtest0-dev libxcb1-dev \
libxcursor1 libxcursor-dev libxft-dev libxft2-dev xfonts-terminus \
build-essential libxrandr-dev libxcb-xfixes0-dev libxpathselect-dev \
libfontconfig1-dev libfreetype6-dev libice-dev libpng-dev \
libpng-tools libpthread-stubs0-dev libsm-dev libxau-dev libxcb-render0-dev \
libxdmcp-dev libxfixes-dev libxrender-dev pkg-config uuid-dev \
x11proto-core-dev x11proto-dev  x11proto-fixes-dev xorg-sgml-doctools \
xtrans-dev zlib1g-dev libxcb-xinput-dev libxcb-xrm-dev libxt-dev \
libxpathselect1.4v5 libxpathselect-dev




build-depend-mint20.sh
# !/BIN/BASH
#
#
#      ___ ___  ___ ____/ /______    ____ _  
#     (_-</ _ \/ -_) __/ __/ __/ |/|/ /  ' \ 
#    /___/ .__/\__/\__/\__/_/  |__,__/_/_/_/ 
#        /_/                                 
#      _                  _                                       _   _           
#     |_)     o |  _| __ | \  _  ._   _  ._   _| __ |\/| o ._ _|_  ) / \    _ |_  
#     |_) |_| | | (_|    |_/ (/_ |_) (/_ | | (_|    |  | | | | |_ /_ \_/ o _> | | 
#                                |                                                
# Script Date: 06/29/2020
# Script Creator: Bruce E. Scott
# Github: https://github.com/bescott9944
# Forums: https://www.ezeelinux.com/talk/bescott9944
#         https://forum.endeavouros.com/bescott9944
# Licensed under the GNU and MIT respectively
#
# PURPOSE Of This Script:
# This Script is for installing/building the new Spectrwm build and the
# dependencies that are needed to build the newer binary version
# of SpectrWm... Doing this build to make a manual install... 06/29/2020 -Bruce
# This script was inspired by a video from The Linux Dabbler YouTube Channel.
# His Channel: https://www.youtube.com/channel/UC3pasXoRUTKPpQaKtlJz62g
# The Video: https://www.youtube.com/watch?v=tYC7qxnjka4
#
# Change Log.
# 07/04/2020
# Made some changes to the dependencies list to work better. -Bruce
#
# NOTE: This Bash file has been limited to 80 characters. You may need edit
# the line and change the "next line break ( \ )" of the bash file to suite
# your needs!
#
# This Install dependencies list has been tested on Mint 20, 06/29/2020
#
# The otherday in a comment in said video, Daniel Olsson said:
# Nice video, sometimes i use sudo apt build-dep "packagename" in this case
# spectrwm, then it will automatically pull in everything you need to
# recompile spectrvm :)
# Wish I known this a few days ago... Lol
# UPDATE: This is not a option at this time. 07/07/2020 -Bruce
#
# Note: Some dependencies require the " -dev " versions of the dependencies and
# other do not. When dependencies are installed they will intern install their
# own dependencies too.. I don't know if this is do to Mint/Ubuntu Repos or not.
# The do the build I needed all these dependencies before the "make" part of the build
# would run with out any error in the output of the "make" and before the binary file
# would compile with out any errors In Linux Mint 20.
#
# The build that I did needed all these dependencies before the "make" part of
# the build would run with out any error in the output and before the binary
# file would compile.
# There was a lot of trial and error an a few hours of testing to produce this
# list. You mileage may vary as dependencies and files change over time.
# with that said this script is a good starting place.
# I have not tested this script on anything other than Linux Mint 20, in a VM
# for testing purpose...
#
# WARNING: USE AT YOUR OWN RISK!
# I or any of my affiliates will or will not be held responsible for any loss
# of data, software, operating systems, damage to, wiping of, altering of, and
# any other thing else that might happen using this script!
# That also includes the urges to drink lots of coffee, Rum, eating of
# junk food, and eating lots of Hot Pockets, and not to limited to the
# building of bonfires, dancing around said bonfires and howling at the MOON!
# YOU HAVE BEEN WARNED! -BES
# sudo apt install g++-multilib g++-9-multilib stterm xfonts-terminus-oblique
# xfonts-cyrillic
#
# Date: 07/09/2020
# Changed the file name to build-depend-mint20.sh
#
#-----------------------------------------------

sudo apt install suckless-tools xterm libxt6 libc6 libx11-6 libx11-xcb1 \
libx11-xcb-dev libxcb-icccm4 libxcb-icccm4-dev libxcb-keysyms1 \
libxcb-keysyms1-dev libxcb-randr0 \
libxcb-randr0-dev libxcb-util1 libxcb-xinput0 libxcb-xtest0-dev libxcb1-dev \
libxcursor1 libxcursor-dev libxft-dev libxft2-dev xfonts-terminus \
build-essential libxrandr-dev libxcb-xfixes0-dev libxpathselect-dev \
libfontconfig1-dev libfreetype6-dev libice-dev libpng-dev \
libpng-tools libpthread-stubs0-dev libsm-dev libxau-dev libxcb-render0-dev \
libxdmcp-dev libxfixes-dev libxrender-dev pkg-config uuid-dev \
x11proto-core-dev x11proto-dev  x11proto-fixes-dev xorg-sgml-doctools \
xtrans-dev zlib1g-dev libxcb-xinput-dev libxcb-xrm-dev libxt-dev \
libxpathselect1.4v5 libxpathselect-dev


